Src/HALALMock/Services/PWM/PWM/PWM.cpp

```diff
@@ -1,110 +1,63 @@
-/*
- * PWM.cpp
- *
- *  Created on: Feb 27, 2023
- *      Author: aleja
- */
-
-#include <PWM/PWM/PWM.hpp>
-#include "ErrorHandler/ErrorHandler.hpp"
-#include "stm32h7xx_hal_def.h"
-
+#include "HALALMock/Services/PWM/PWM/PWM.hpp"
 PWM::PWM(Pin& pin) {
-	if (not TimerPeripheral::available_pwm.contains(pin)) {
-
-		ErrorHandler("Pin %s is not registered as an available PWM", pin.to_string());
-		return;
-	}
-
-	TimerPeripheral& timer = TimerPeripheral::available_pwm.at(pin).first;
-	TimerPeripheral::PWMData& pwm_data = TimerPeripheral::available_pwm.at(pin).second;
-
-	if (pwm_data.mode != TimerPeripheral::PWM_MODE::NORMAL) {
-		ErrorHandler("Pin %s is not registered as a NORMAL PWM", pin.to_string());
+	/*By HalalMock we only will test the logical interface of PWM
+	if the PWM works, however when you try in your micro doesn't work,
+	you should check the timers and channels, that could be the problem 
+	*/
+	EmulatedPin &pin_memory=SharedMemory::get_pin(pin);
+	if(pin_memory.type==PinType::NOT_USED){
+		pin_memory.type=PinType::PWM;
+		pin_memory.PinData.PWM.duty_cycle=0.0f;
+		pin_memory.PinData.PWM.is_on=false;
+		//let's point our class variables to the variables from PinModel
+		this->duty_cycle=&(pin_memory.PinData.PWM.duty_cycle);
+		this->frequency=&(pin_memory.PinData.PWM.frequency);
+		this->is_on=&(pin_memory.PinData.PWM.is_on);
+		//default values
+		*(this->duty_cycle)=0.0f;
+		*(this->is_on)=false;
+		*(this->frequency)=0;
+	}else{
+		ErrorHandler("Pin %s is being used already",pin.to_string());
 	}
-
-	peripheral = &timer;
-	channel = pwm_data.channel;
-
-	Pin::inscribe(pin, TIMER_ALTERNATE_FUNCTION);
-	timer.init_data.pwm_channels.push_back(pwm_data);
-
-	duty_cycle = 0;
-	is_initialized = true;
 }
 
 void PWM::turn_on() {
-  if(not is_initialized){
-	ErrorHandler("PWM was not initialized");
-  }
-  if(is_on) return;
-  if(HAL_TIM_PWM_Start(peripheral->handle, channel) != HAL_OK) {
-    ErrorHandler("PWM did not turn on correctly", 0);
-  }
-  is_on = true;
+	if(*(this->frequency)==0){
+		ErrorHandler("You need to set a frequency before turn_on the PWM :%d",*(this->frequency));
```

Comment:
If you don't check this condition in set_frequency I wouldn't check it here either, if there is the need for the frequency to be !=0 then the check should be everywhere, otherwise it should be nowhere, as the condition can be easily broken by the user code once started

Reply:
Ok true